model_api/python/model_api/models/visual_prompting.py

```diff
@@ -0,0 +1,784 @@
+        the features are not presented in the internal object state.
+        """
+        if self.has_reference_features():
+            return VisualPromptingFeatures(
+                np.copy(self._reference_features), np.copy(self._used_indices)
+            )
+
+        raise RuntimeError("Reference features are not generated")
+
+    def learn(
+        self,
+        image: np.ndarray,
+        boxes: list[Prompt] | None = None,
+        points: list[Prompt] | None = None,
+        polygons: list[Prompt] | None = None,
+        reset_features: bool = False,
+    ) -> tuple[VisualPromptingFeatures, np.ndarray]:
+        """
+        Executes `learn` stage of SAM ZSL pipeline.
+
+        Reference features are updated according to newly arrived prompts.
+        Features corresponding to the same labels are overridden during
+        consequent learn() calls.
+
+        Args:
+            image (np.ndarray): HWC-shaped image
+            boxes (list[Prompt] | None, optional): Prompts containing bounding boxes (in XYXY torchvision format)
+              and their labels (ints, one per box). Defaults to None.
+            points (list[Prompt] | None, optional): Prompts containing points (in XY format)
+              and their labels (ints, one per point). Defaults to None.
+            polygons: (list[Prompt] | None): Prompts containing polygons (a sequence of points in XY format)
+              and their labels (ints, one per polygon).
+              Polygon prompts are used to mask out the source features without implying decoder usage. Defaults to None.
+            reset_features (bool, optional): Forces learning from scratch. Defaults to False.
+
+        Returns:
+            tuple[VisualPromptingFeatures, np.ndarray]: return values are the updated VPT reference features and reference masks.
+            The shape of the reference mask is N_labels x H x W, where H and W are the same as in the input image.
+        """
+
+        if boxes is None and points is None and polygons is None:
+            raise RuntimeError(
+                "boxes, polygons or points prompts are required for learning"
+            )
+
+        if reset_features or not self.has_reference_features():
+            self.reset_reference_info()
+
+        processed_prompts = self.decoder.preprocess(
+            {
+                "bboxes": [box.data for box in boxes] if boxes else None,
+                "points": [point.data for point in points] if points else None,
+                "labels": {
+                    "bboxes": [box.label for box in boxes] if boxes else None,
+                    "points": [point.label for point in points] if points else None,
+                },
+                "orig_size": image.shape[:2],
+            },
+        )
+
+        if polygons is not None:
+            for poly in polygons:
+                processed_prompts.append({"polygon": poly.data, "label": poly.label})
```

Comment:
I think the additional workflow is needed to avoid getting duplicated masks:
check if there is bbox or point prompt that targets the same object with polygon -> if so, remove corresponding bbox or point
What do you think?

Reply:
I think we should discuss this with Max. I'm not sure if prompts cleaning should be a responsibility of the ZSL pipeline, since the annotation engine is somewhat external.

Reply:
What happens if we have duplicate prompts? And if they have different labels?

Reply:
> I think we should discuss this with Max. I'm not sure if prompts cleaning should be a responsibility of the ZSL pipeline, since the annotation engine is somewhat external.

Ah, I agree with your opinion. Let's pass it.

Reply:
> What happens if we have duplicate prompts? And if they have different labels?

No problem because `learn` just outputs reference features corresponding to given prompts.
But there are reference features that target the same objects and duplicated masks will be generated.
For this case, `_inspect_overlapping_areas` will filter those that have the same label.
https://github.com/sovrasov/model_api/blob/d6c9ad0a0ea24b9033c3d7637e239c27bba473fe/model_api/python/model_api/models/visual_prompting.py#L737-L741
But masks with different labels will not be applicable, resulting in duplicate results.

Reply:
Thanks for the explanation. I agree that it's user's responsibility for processing the prompts.